homeassistant/components/zha/light.py

```diff
@@ -224,17 +228,51 @@ async def async_turn_on(self, **kwargs):
         effect = kwargs.get(light.ATTR_EFFECT)
         flash = kwargs.get(light.ATTR_FLASH)
 
-        if brightness is None and self._off_brightness is not None:
+        # If the light is currently off but a turn_on call with a color/temperature is sent,
+        # the light needs to be turned on first at a low brightness level where the light is immediately transitioned
+        # to the correct color. Afterwards, the transition is only from the low brightness to the new brightness.
+        # Otherwise, the transition is from the color the light had before being turned on to the new color.
+        # This can look especially bad with transitions longer than a second.
+        color_provided_from_off = (
+            not self._state
+            and brightness_supported(self._attr_supported_color_modes)
+            and (light.ATTR_COLOR_TEMP in kwargs or light.ATTR_HS_COLOR in kwargs)
+        )
+        final_duration = duration
+        if color_provided_from_off:
+            # Set the duration for the color changing commands to 0.
+            duration = 0
+
+        if (
+            brightness is None
+            and (self._off_with_transition or color_provided_from_off)
+            and self._off_brightness is not None
+        ):
             brightness = self._off_brightness
 
+        if brightness is not None:
+            level = min(254, brightness)
+        else:
+            level = self._brightness or 254
```

Comment:
Can self._brightness be 0 here? If so, is this going to do what you expect?

Reply:
Looks like this already existed below. Wonder if it works correctly…

Reply:
If ``self._brightness`` was somehow ``0`` and there's no given ``brightness`` using the service call, then I'd expect the light to turn to full brightness (254) if a transition is specified. I can't think of any case where ``self._brightness`` is used afterwards when it's ``0`` in the first place though.

That code seems correct to me (and seems to work).

(I just thought of another thing with ``off_brightness`` being overridden when the light is already off though -> will make a comment for that)